subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformEnvironmentIntegrationTest.groovy

```diff
@@ -0,0 +1,242 @@
+
+/**
+ * Tests the state of the application classpath in the forked test process to ensure the correct
+ * test framework dependencies are exposed to the user's test code. Additionally tests environmental
+ * state like system properties, and environment variables.
+ *
+ * <p>This test intentionally does not extend {@link JUnitPlatformIntegrationSpec} in order to have
+ * complete control over the configuration of the test setup</p>
+ */
+class JUnitPlatformEnvironmentIntegrationTest extends AbstractIntegrationSpec {
+
+    // The versions tested against here are intentionally different than the version of junit-platform-launcher
+    // that Gradle will load from the distribution. This way, we can use the version on the application classpath
+    // to determine whether the launcher was loaded from the distribution or from the test runtime classpath.
+    private static final JUNIT_JUPITER_VERSION = '5.8.1'
+    private static final JUNIT_PLATFORM_VERSION = '1.8.1'
+    private static final OPENTEST4J_VERSION = '1.2.0'
+
+    def setup() {
+        buildFile << """
+            plugins {
+                id 'java-library'
+            }
+
+            ${mavenCentralRepository()}
+
+            test {
+                useJUnitPlatform()
+
+                systemProperties.isJava9 = "\${JavaVersion.current().isJava9Compatible()}"
+                systemProperties.testSysProperty = 'value'
+                systemProperties.projectDir = projectDir
+                environment.TEST_ENV_VAR = 'value'
+                testLogging.showStandardStreams = true
+            }
+        """
+    }
+
+    def "should prompt user to add dependencies when they are not in test runtime classpath"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                compileOnly 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+            }
+        """
+        file('src/test/java/org/example/ExampleTest.java') << """
+            package org.example;
+            public class ExampleTest {
+                @org.junit.jupiter.api.Test
+                public void ok() { }
+            }
+        """
+
+        when:
+        fails('test')
+
+        then:
+        new DefaultTestExecutionResult(testDirectory)
+            .testClassStartsWith('Gradle Test Executor')
+            .assertExecutionFailedWithCause(containsString('consider adding an engine implementation JAR to the classpath'))
+    }
+
+    def "automatically loads junit-platform-launcher from distribution"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+            }
+        """
+
+        addClasspathTest("""
+            assertEquals(jars.get(0), "junit-jupiter-params-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(1), "junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
```

Comment:
❕ Checking the exact order of the test classpath (beyond the first 2 gradle-worker and test jar), seems brittle to me.  What guarantee is there that this order won't change (to alphabetical, for instance; or when a newer version of junit's dependency graph changes)?  I think this would be better as just a "contains" check.

Reply:
I don't think there is any guarantee of these orderings. I'll switch to a contains check